### PR TITLE
feat: Adds support for ML-DSA keys and certificates

### PR DESCRIPTION
**Pull request by jobybejoy**
Summary
Implements support for ML-DSA (Module-Lattice-Based Digital Signature Algorithm) keys and certificates in the AWS IAM Roles Anywhere Credential Helper, supporting all three NIST FIPS 204 parameter sets: ML-DSA-44, ML-DSA-65, and ML-DSA-87.

Changes
• New ML-DSA utilities (mldsa_utils.go): Core ML-DSA key handling, OID definitions, and signature operations
• Enhanced PKCS#8 support (pkcs8_utils.go): Extended private key parsing for ML-DSA formats
• Updated signer logic (signer.go): Integrated ML-DSA signature algorithms into existing signing workflow
• File system signer (file_system_signer.go): Added ML-DSA key detection and loading
• Comprehensive tests (mldsa_utils_test.go): Full test coverage for all ML-DSA variants
• Dependencies: Added Cloudflare CIRCL library for ML-DSA implementation

Technical Details
• Supports NIST FIPS 204 standard ML-DSA parameter sets (44, 65, 87)
• Implements proper ASN.1 OID handling for ML-DSA algorithms
• Maintains backward compatibility with existing RSA/ECDSA workflows
• Uses widely used cryptographic library Cloudflare CIRCL to sign using ML-DSA keys. 

Testing
• Key generation, parsing, and signature verification tests
• Error handling and edge case validation

This enhancement enables post-quantum cryptographic signatures for IAM Roles Anywhere authentication, preparing for future quantum-resistant security requirements.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Supersedes #155

### CODE REVIEW

File: aws_signing_helper/mldsa_utils.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+	"os"
+
+	"github.com/cloudflare/circl/sign/mldsa/mldsa44"
+	"github.com/cloudflare/circl/sign/mldsa/mldsa65"
+	"github.com/cloudflare/circl/sign/mldsa/mldsa87"
+)
+
+// MLDSA OIDs as defined in NIST FIPS 204
+var (
+	oidMLDSA44 = asn1.ObjectIdentifier{2, 16, 840, 1, 101, 3, 4, 3, 17} // id-ml-dsa-44
```
[the remark this answers is not in the document]
Comment (jobybejoy) [reply aws_signing_helper/mldsa_utils.go]: Ack! likely introduced in a iteration - I've consolidated the OID's into mldsa_utils.go. I guess eventually when there is native support for ML-DSA in Go. A bunch these blocks can be removed.

File: aws_signing_helper/signer.go [hunk 2]
```diff
@@ -800,6 +902,34 @@ func ReadPKCS12Data(certificateId string) (certChain []*x509.Certificate, privat
 	return certChain, privateKey, nil
 }
 
+// isMLDSACertificate checks if a certificate uses ML-DSA by examining its public key algorithm OID
+func isMLDSACertificate(cert *x509.Certificate) (bool, string, error) {
+	// Parse the certificate's SubjectPublicKeyInfo
+	var spki SubjectPublicKeyInfo
```
[the remark this answers is not in the document]
Comment (jobybejoy) [reply aws_signing_helper/signer.go]: Will be moving this function into mldsa_utils.go

File: generate-credential-process-data.sh [hunk 3]
```diff
@@ -2,13 +2,18 @@
 
 # Simple script to generate a CA certificate/private key
 # and end-entity certificate/private key for use with 
-# Roles Anywhere
+# Roles Anywhere, plus ML-DSA test fixtures
```
Comment (camerondugan) [inline generate-credential-process-data.sh]: Nitpick: I could be wrong, but I though that the makefile was responsible for test fixture generation? Could we copy/paste this new logic into a recipe and make it required for the test targets that use these mldsa artifacts?

File: aws_signing_helper/mldsa_utils_test.go [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+	// Benchmark-style test to measure performance of encrypted ML-DSA key operations
+	if testing.Short() {
+		t.Skip("Skipping benchmark test in short mode")
+	}
+
+	keyFile := "../tst/certs/mldsa44-key-pkcs8-aes256cbc.pem"
+	
+	// Check if fixture exists
+	if _, err := os.Stat(keyFile); os.IsNotExist(err) {
+		t.Skip("Skipping benchmark - ML-DSA fixture not available")
```
Comment (camerondugan) [inline aws_signing_helper/mldsa_utils_test.go]: We should have build flag or env var that enables hard enforcement for expecting mldsa tests to run, currently with these changes it's possible to make a ci/cd change that accidentaly removes generation of fixtures and we would not know because these tests will just pass/skip.

File: aws_signing_helper/signer_test.go [hunk 5]
```diff
@@ -791,3 +798,151 @@ func GetMockedCreateSessionResponseServer() *httptest.Server {
 		  }`))
 	}))
 }
+func TestMLDSAEncryptedKeySigningWorkflow(t *testing.T) {
```
Comment (camerondugan) [inline aws_signing_helper/signer_test.go]: There currently is no unit test that guarantees the delegation from FileSystemSigner.Sign() to the underlying MLDSA key's Sign() function. This test could be extended a bit to sign some data and verify against the cert to cover this case.

File: aws_signing_helper/mldsa_utils.go [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+
+// PKCS#8 structure for parsing private keys
+type pkcs8PrivateKey struct {
+	Version    int
+	Algo       pkix.AlgorithmIdentifier
+	PrivateKey []byte
+}
+
+// readMLDSAPrivateKey attempts to read an MLDSA private key from a PEM file
+func readMLDSAPrivateKey(privateKeyId string) (MLDSAPrivateKey, error) {
```
Comment (030emily) [inline aws_signing_helper/mldsa_utils.go]: This could be simplified by calling the `ParseMLDSAFromPKCS8` function

File: aws_signing_helper/file_system_signer.go [hunk 7]
```diff
@@ -35,13 +35,28 @@ func (fileSystemSigner *FileSystemSigner) Public() crypto.PublicKey {
...
 
 func (fileSystemSigner *FileSystemSigner) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) (signature []byte, err error) {
 	privateKey, _, _ := fileSystemSigner.readCertFiles()
+
+	// Check if it's an MLDSA key - MLDSA signs the message directly without hashing
+	mldsakey, ok := privateKey.(MLDSAPrivateKey)
+	if ok {
+		// MLDSA signs the digest directly (no additional hashing)
+		sig := mldsakey.Sign(digest, nil)
+		return sig, nil
```
[the remark this answers is not in the document]
Comment (jobybejoy) [reply aws_signing_helper/file_system_signer.go]: Based on running a few tests something is populating the opts.HashFunc() its unclear to me at the moment where its being populated. I will have to revisit this comment at a later time. 

```
2026/01/21 11:26:10 ML-DSA Sign called: opts.HashFunc() = SHA-256
2026/01/21 11:26:10 ML-DSA Sign called: opts.HashFunc() = SHA-256
2026/01/21 11:26:10 ML-DSA Sign called: opts.HashFunc() = SHA-256
```

File: aws_signing_helper/mldsa_utils_test.go (hunk 4 above)
Comment (jobybejoy) [reply aws_signing_helper/mldsa_utils_test.go]: Let me know if you spot more of these. I would like to remove them.

File: generate-credential-process-data.sh (hunk 3 above)
Comment (jobybejoy) [reply generate-credential-process-data.sh]: Changes in this file might be irrelevant; I was unaware that fixtures were being built in the make file.

Comment (jobybejoy) [reply generate-credential-process-data.sh]: Created a new phony target for `mldsa-fixtures` in https://github.com/aws/rolesanywhere-credential-helper/pull/157/commits/2132951dcd9e1f79578a83bdc5fe172fca3efd0c and moved the fixtures into `tst/mldsa-fixtures`. In addition ML-DSA private keys and certificate are also included in the commit.
